enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+```bash
+oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/rbac.yaml  
+oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/aws-cloud-controller-manager-daemonset.yaml
+```
+
+##### Azure
+
+Main repository: https://github.com/kubernetes-sigs/cloud-provider-azure
+Sample manifests: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/examples/out-of-tree
+
+##### GCP
+
+Main repository: https://github.com/kubernetes/cloud-provider-gcp
+Sample manifests: https://github.com/kubernetes/cloud-provider-gcp/tree/master/deploy
+
+##### vSphere
+
+Main repository: https://github.com/kubernetes/cloud-provider-vsphere
+Sample manifests: https://github.com/kubernetes/cloud-provider-vsphere/tree/master/manifests/controller-manager
+
+## Upgrade/Downgrade strategy
+
+While a set of `cloud-controller-manager` pods is running under a `DaemonSet` provisioned by the operator, aquiring leadership over cloud specific control loops in a HA openshift deployment will be managed by a cross-namespace leader-election migration.
+
+Upstream [proposal](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md#motivation) describes the main implementaion details for this component. Their proposed implementation for cross-namespace leadership delegation would help Openshift to preserve HA during upgrage from in-tree to the out-of-tree.
+
+In case the cluster does not require to be HA, without integrating this feature an upgrade under bare CVO management will lead to two simultanious replicas of `cloud-controllers` running at the same time at some moment.
+
+### Test Plan
+
+- Each provider at the time of migration should have a working CI implemetation, which will assist in testing the provisioning with out-of-tree support enabled.
+- Ensure that transition between the in-tree and the out-of-tree will be handled by the `Infrastracture` `status.platformStatus.external` field being set.
+- Ensure the in-tree and the out-of-tree providers could co-exist in the cluster if this is supported.
+- Ensure that upgrade from a release running on the in-tree provider only to the out-of-tree succeeds.
+
+## Other notes
+
+##### Removing a deprecated feature
+
+Upstream is currently planning to remove support for the in-tree providers with OCP-4.8 (kubernetes 1.21) - based on the slack [conversation](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600). The enchancement [describes](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md) steps towards this process. Our goal is to have a working alternative for the in-tree providers ready to be used on-par with old implementation befor the upstream release will remove this feature from core Kubernetes repository.
+
+## Timeline
```

Comment:
this is expecte to have a timeline for our implementation, not for upstream. Please add, especially answering questions around https://github.com/openshift/enhancements/pull/463/files#r492579924.